pkg/operator/controller/dns_status_test.go

```diff
@@ -0,0 +1,354 @@
+			availStatus:    operatorv1.ConditionFalse,
+		},
+		{
+			description:    "no cluster ip, 2/2 pods available",
+			clusterIP:      "",
+			avail:          2,
+			desire:         2,
+			degradeStatus:  operatorv1.ConditionTrue,
+			progressStatus: operatorv1.ConditionFalse,
+			availStatus:    operatorv1.ConditionFalse,
+		},
+		{
+			description:    "daemonset pod available on 0/0 nodes",
+			clusterIP:      "1.2.3.4",
+			avail:          0,
+			desire:         0,
+			degradeStatus:  operatorv1.ConditionTrue,
+			progressStatus: operatorv1.ConditionFalse,
+			availStatus:    operatorv1.ConditionFalse,
+		},
+		{
+			description:    "daemonset pod available on 0/2 nodes",
+			clusterIP:      "1.2.3.4",
+			avail:          0,
+			desire:         2,
+			degradeStatus:  operatorv1.ConditionTrue,
+			progressStatus: operatorv1.ConditionTrue,
+			availStatus:    operatorv1.ConditionFalse,
+		},
+		{
+			description:    "daemonset pod available on 1/2 nodes",
+			clusterIP:      "1.2.3.4",
+			avail:          1,
+			desire:         2,
+			degradeStatus:  operatorv1.ConditionTrue,
+			progressStatus: operatorv1.ConditionTrue,
+			availStatus:    operatorv1.ConditionTrue,
+		},
+		{
+			description:    "daemonset pod available on 2/2 nodes",
+			clusterIP:      "1.2.3.4",
+			avail:          2,
+			desire:         2,
+			degradeStatus:  operatorv1.ConditionFalse,
+			progressStatus: operatorv1.ConditionFalse,
+			availStatus:    operatorv1.ConditionTrue,
+		},
+	}
+
+	for i, tc := range testCases {
+		ds := &appsv1.DaemonSet{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: fmt.Sprintf("dns-%d", i+1),
+			},
+			Status: appsv1.DaemonSetStatus{
+				DesiredNumberScheduled: tc.desire,
+				NumberAvailable:        tc.avail,
+			},
+		}
+
+		expected := []operatorv1.OperatorCondition{
+			{
+				Type:   operatorv1.OperatorStatusTypeDegraded,
+				Status: tc.degradeStatus,
+			}, {
+				Type:   operatorv1.OperatorStatusTypeAvailable,
+				Status: tc.availStatus,
+			},
+			{
+				Type:   operatorv1.OperatorStatusTypeProgressing,
+				Status: tc.progressStatus,
+			},
+		}
+		actual := computeDNSStatusConditions([]operatorv1.OperatorCondition{}, tc.clusterIP, ds)
+		gotExpected := true
+		if len(actual) != len(expected) {
+			gotExpected = false
+		}
+		for _, conditionA := range actual {
+			foundMatchingCondition := false
+
+			for _, conditionB := range expected {
+				if conditionA.Type == conditionB.Type &&
+					conditionA.Status == conditionB.Status {
+					foundMatchingCondition = true
+					break
+				}
+			}
+
+			if !foundMatchingCondition {
+				gotExpected = false
+			}
+		}
+		if !gotExpected {
+			t.Fatalf("%q: expected %#v, got %#v", tc.description,
+				expected, actual)
+		}
+	}
+}
+
+func TestSetDNSStatusCondition(t *testing.T) {
+	testCases := []struct {
+		description   string
+		oldConditions []operatorv1.OperatorCondition
+		newCondition  *operatorv1.OperatorCondition
+		expected      []operatorv1.OperatorCondition
```

Comment:
Instead of using operatorv1.OperatorCondition, we could use conditions{available: true}, conditions{degraded: false, progressing: true} and build the actual conditions based on this concise input before calling the test method.